src/main/java/uk/gov/pay/directdebit/mandate/services/gocardless/GoCardlessMandateStateCalculator.java

```diff
@@ -51,14 +93,37 @@
                             GOCARDLESS_ACTIONS_THAT_CHANGE_STATE);
                 })
                 .orElse(Optional.empty());
+    }
+
+    private DirectDebitStateWithDetails<MandateState> mapEventToState(Event event, Mandate mandate) {
+        if (event instanceof GoCardlessEvent) {
+            return mapGoCardlessEventToState((GoCardlessEvent) event, mandate);
+        } else if (event instanceof GovUkPayEvent) {
+            return mapGovUkPayEventToState((GovUkPayEvent) event, mandate);
+        } else {
+            throw new IllegalArgumentException(format("Unexpected Event of type %s", event.getClass()));
+        }
+    }
+
+    private DirectDebitStateWithDetails<MandateState> mapGoCardlessEventToState(GoCardlessEvent goCardlessEvent, Mandate mandate) {
+        if (!GOCARDLESS_ACTION_TO_MANDATE_STATE.containsKey(goCardlessEvent.getAction())) {
+            throw new UnexpectedGoCardlessEventActionException(goCardlessEvent.getAction(), mandate);
```

Comment:
Can this ever reasonably happen given that we got the relevant event based on the key set of the map?

Reply:
no, not reasonably, should I just let it throw a NullPointerException if somehow the DAO returns something unexpected?

Reply:
I’d say so. It’s still a `RuntimeException`, so the effect will be basically the same.